File: Core/src/main/java/jokatu/game/event/dialog/BaseDialogRequestBuilder.java
package jokatu.game.event.dialog;

import jokatu.game.input.Input;
import jokatu.game.player.Player;
import jokatu.ui.Form;
import ophelia.function.ExceptionalBiConsumer;
import org.jetbrains.annotations.NotNull;

class BaseDialogRequestBuilder<P extends Player, I extends Input> implements MainDialogRequestBuilder<P, I> {
	private final Class<I> expectedInput;
	private final P player;
	private final String title;
	private final String message;
	private final ExceptionalBiConsumer<? super I, ? super P, ?> consumer;

	private Form form;

	BaseDialogRequestBuilder(
			@NotNull Class<I> expectedInput,
			@NotNull P player,
			@NotNull String title,
			@NotNull String message,
			@NotNull ExceptionalBiConsumer<? super I, ? super P, ?> consumer
	) {
		this.expectedInput = expectedInput;
		this.player = player;
		this.title = title;
		this.message = message;
		this.consumer = consumer;
	}

	@Override
	@NotNull
	public DialogRequest<P, I> build() {
		return new DialogRequest<>(expectedInput, player, title, message, consumer, form);
	}

	@Override
	@NotNull
	public MainDialogRequestBuilder<P, I> noöp() {
		return this;
	}

	@NotNull
	@Override
	public MainDialogRequestBuilder<P, I> withForm(@NotNull Form form) {
		this.form = form;
		return this;
	}
}

File: Core/src/main/kotlin/jokatu/game/event/dialog/BaseDialogRequestBuilder.kt
package jokatu.game.event.dialog

import jokatu.game.input.Input
import jokatu.game.player.Player
import jokatu.ui.Form
import ophelia.function.ExceptionalBiConsumer

internal class BaseDialogRequestBuilder<P : Player, I : Input>(
		private val expectedInput: Class<I>,
		private val player: P,
		private val title: String,
		private val message: String,
		private val consumer: ExceptionalBiConsumer<in I, in P, *>
) : MainDialogRequestBuilder<P, I> {

	private var form: Form? = null
	private var cancellable = true

	override fun build(): DialogRequest<P, I> {
		return DialogRequest(expectedInput, player, title, message, consumer, form, cancellable)
	}

	override fun noöp(): MainDialogRequestBuilder<P, I> {
		return this
	}

	override fun withForm(form: Form): MainDialogRequestBuilder<P, I> {
		this.form = form
		return this
	}

	override fun withoutCancelButton(): MainDialogRequestBuilder<P, I> {
		this.cancellable = false
		return this
	}
}
